plugins/inputs/intel_powerstat/services/rapl.go

```diff
@@ -0,0 +1,274 @@
+			r.data[socketID] = &data.RaplData{
+				SocketCurrentEnergy: 0,
+				DramCurrentEnergy:   0,
+				SocketEnergy:        0,
+				DramEnergy:          0,
+				ReadDate:            0,
+			}
+		}
+	}
+}
+
+func (r *RaplServiceImpl) findDramFolders() {
+	intelRaplPrefix := "intel-rapl:"
+	intelRaplDram := fmt.Sprintf("%s%s", intelRaplPrefix, "[0-9]*[0-9]*")
+	// Clean existing map
+	r.dramFolders = make(map[string]string)
+
+	for socketID := range r.data {
+		path := fmt.Sprintf(IntelRaplSocketPartialPath, IntelRaplPath, socketID)
+		pathsToRaplFolders, err := r.fs.GetStringsMatchingPatternOnPath(fmt.Sprintf("%s/%s", path, intelRaplDram))
+		if err != nil {
+			r.log.Errorf("error during lookup for rapl dram: %v", err)
+			continue
+		}
+		raplFolders := make([]string, 0)
+		for _, folderPath := range pathsToRaplFolders {
+			raplFolders = append(raplFolders, filepath.Base(folderPath))
+		}
+
+		r.findDramFolder(raplFolders, socketID)
+	}
+}
+
+func (r *RaplServiceImpl) calculateData(socketID string, energyFile io.Reader, dramEnergyFile io.Reader) error {
+	newSocketEnergy, _, err := r.readEnergyInJoules(energyFile)
+	if err != nil {
+		return err
+	}
+
+	newDramEnergy, readDate, err := r.readEnergyInJoules(dramEnergyFile)
+	if err != nil {
+		return err
+	}
+
+	interval := ConvertNanoSecondsToSeconds(readDate - r.data[socketID].ReadDate)
+	r.data[socketID].ReadDate = readDate
+
+	if newSocketEnergy > r.data[socketID].SocketEnergy {
+		r.data[socketID].SocketCurrentEnergy = (newSocketEnergy - r.data[socketID].SocketEnergy) / interval
+	} else {
+		maxEnergy, err := r.GetMaxEnergyRangeUj(socketID)
+		if err != nil {
+			return err
+		}
+		// When energy_uj counter reaches maximum value defined in max_energy_range_uj file it
+		// starts counting from 0.
+		r.data[socketID].SocketCurrentEnergy = maxEnergy - r.data[socketID].SocketEnergy + newSocketEnergy
+	}
+
+	if newDramEnergy > r.data[socketID].DramEnergy {
+		r.data[socketID].DramCurrentEnergy = (newDramEnergy - r.data[socketID].DramEnergy) / interval
+	} else {
+		dramMaxEnergy, err := r.GetDramMaxEnergyRangeUj(socketID)
+		if err != nil {
+			return err
+		}
+		// When dram energy_uj reaches maximum value defined in max_energy_range_uj file it
+		// starts counting from 0.
+		r.data[socketID].DramCurrentEnergy = dramMaxEnergy - r.data[socketID].DramEnergy + newDramEnergy
+	}
+	r.data[socketID].SocketEnergy = newSocketEnergy
+	r.data[socketID].DramEnergy = newDramEnergy
+
+	return nil
+}
+
+func (r *RaplServiceImpl) findDramFolder(raplFolders []string, socketID string) {
+	for _, raplFolder := range raplFolders {
+		potentialDramPath := fmt.Sprintf(IntelRaplDramPartialPath, IntelRaplPath, socketID, raplFolder)
+		nameFilePath := fmt.Sprintf(IntelRaplDramNamePartialPath, potentialDramPath)
+		read, err := r.fs.ReadFile(nameFilePath)
+		if err != nil {
+			r.log.Errorf("error reading file on path: %s, err: %v", nameFilePath, err)
+			continue
+		}
```

Comment:
Should we really continue here of fail? Is there a real use-case (hot-plugging CPU?) where we want to go on and accept the error?

Reply:
Sure: hot-plugging RAPL system (when RAPL's paths appear and disappear in filesystem).
```
sudo modprobe intel_rapl
sudo modprobe -r intel_rapl
```